substrate/runtime/contract/src/lib.rs

```diff
@@ -142,22 +181,94 @@ impl<'a, T: Ext + 'a> Runtime<'a, T> {
 			}
 		}
 	}
+	/// Returns how much gas left from the initial budget.
+	fn gas_left(&self) -> u64 {
+		self.gas_limit
+			.checked_sub(self.gas_used)
+			.expect(
+				"gas_used is always incremented via Self::charge;
+				Self::charge ensures that gas_used is always less than or equal to gas_limit;
+				this substraction can never underflow;
+				qed;
+				"
+			)
+	}
+}
+
+fn to_execution_result<T: Ext>(
+	runtime: Runtime<T>,
+	run_err: Option<sandbox::Error>,
+) -> Result<ExecutionResult, Error> {
+	let mut return_data = Vec::new();
+
+	// Check the exact type of the error. It could be plain trap or
+	// special runtime trap the we must recognize.
+	match (run_err, runtime.special_trap) {
+		// No traps were generated. Proceed normally.
+		(None, None) => {},
+		// Special case. The trap was the result of the execution `return` host function.
+		(Some(sandbox::Error::Execution), Some(SpecialTrap::Return(rd))) => {
+			return_data = rd
+		}
+		// Any other kind of a trap should result in a failure.
+		(Some(_), _) => {
+			return Err(Error::Invoke);
+		}
+		_ => {
+			// All possible cases should have been handled.
+			// If the control flow reached here, then it is a logic error.
+			unreachable!();
+		}
+	}
+
+	Ok(ExecutionResult {
+		gas_left: runtime.gas_meter.gas_left(),
+		return_data,
+	})
+}
+
+/// The result of execution of a smart-contract.
+#[derive(PartialEq, Eq)]
+#[cfg_attr(feature = "std", derive(Debug))]
+pub struct ExecutionResult {
+	return_data: Vec<u8>,
+	gas_left: u64,
+}
+
+impl ExecutionResult {
+	/// The result produced by the execution of the contract.
+	///
+	/// The contract can designate some buffer at the execution time via a special function.
+	/// If contract called this function with non-empty buffer it will be copied here.
+	///
+	/// Note that gas is already charged for returning the data.
+	pub fn return_data(&self) -> &[u8] {
+		&self.return_data
+	}
+
+	/// How much gas left after the execution of the contract.
+	pub fn gas_left(&self) -> u64 {
+		self.gas_left
+	}
 }
 
 /// Execute the given code as a contract.
 pub fn execute<'a, T: Ext>(
 	code: &[u8],
 	ext: &'a mut T,
 	gas_limit: u64,
-) -> Result<(), Error> {
+) -> Result<ExecutionResult, Error> {
 	// ext_gas(amount: u32)
 	//
 	// Account for used gas. Traps if gas used is greater than gas limit.
 	//
 	// - amount: How much gas is used.
-	fn ext_gas<T: Ext>(e: &mut Runtime<T>, args: &[sandbox::TypedValue]) -> Result<sandbox::ReturnValue, sandbox::HostError> {
+	fn ext_gas<T: Ext>(
+		e: &mut Runtime<T>,
+		args: &[sandbox::TypedValue],
+	) -> Result<sandbox::ReturnValue, sandbox::HostError> {
 		let amount = args[0].as_i32().unwrap() as u32;
```

Comment:
This might be out of scope, but i'm unsure about those unwraps

I know they might be guaranteed by macro syntax atm, but any bug in upstream library might lead to a debugging hell

Reply:
You're totally right, this is something we need to change eventually! 

That's on my todo list, however it's kinda blocked on the design of JS side.